### ISSUE TITLE
Using Firebase in XCTest when Host application is set to None

### ISSUE THREAD

**Issue by ugenlik** [firebase/firebase-ios-sdk#146]
* Xcode version: 8.3.2
  * Firebase SDK version: 4.0
  * Library version: 4.0
  * Firebase Product: core

My unit tests don't use a host application. 

But I wanna conduct some unit tests and I need to include Firebase in my test target.

`GoogleService-Info.plist` is added to both actual target and unit tests. 

Because I couldn't mock the Firebase I just wanna call `FirebaseApp.configure()` in my principal test class. 

However, it says it initialized, it throws a stack error; 
```
    import Firebase
    
    class TestSetup: NSObject {
        override init() {
            super.init()
            
            //let url = Bundle(for: type(of: self)).path(forResource: "GoogleService-Info", ofType: "plist")
            //let bundleOption:FirebaseOptions = FirebaseOptions.init(contentsOfFile: url!)!
            
            //let bundleOption:FirebaseOptions = FirebaseOptions.defaultOptions()!
            //bundleOption.bundleID = "com.xxxx.HuginTests"
            //FirebaseApp.configure(options: bundleOption)
            
            FirebaseApp.configure()
            //let _ = FirebaseConfig.sharedInstance
            
        }
    }
```

It looks like firebase is `Successfully initialized`  but this throws the following error:

```
    2017-07-19 15:37:42.000 xctest[2990] <Warning> [Firebase/Analytics][I-ACS005000] The AdSupport Framework is not currently linked. Some features will not function properly. Learn more at 
    2017-07-19 15:37:42.023 xctest[2990] <Notice> [Firebase/Crash][I-CRA000004] Successfully initialized
    2017-07-19 15:38:00.141 xctest[2990:122374] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSPlaceholderArray initWithObjects:count:]: attempt to insert nil object from objects[0]'
    *** First throw call stack:
    (
    	0   CoreFoundation                      0x00000001049a5b0b __exceptionPreprocess + 171
    	1   libobjc.A.dylib                     0x000000010440a141 objc_exception_throw + 48
    	2   CoreFoundation                      0x00000001048ba56d -[__NSPlaceholderArray initWithObjects:count:] + 269
    	3   CoreFoundation                      0x00000001048c12b4 +[NSArray arrayWithObjects:count:] + 52
    	4   HuginTests                          0x0000000113bccdbf -[RCNConfigDBManager loadMainTableWithBundleIdentifier:fromSource:] + 163
    	5   HuginTests                          0x0000000113bccc8e __62-[RCNConfigDBManager loadMainWithBundleIdentifier:fromSource:]_block_invoke + 34
    	6   libdispatch.dylib                   0x0000000105f33792 _dispatch_client_callout + 8
    	7   libdispatch.dylib                   0x0000000105f183de _dispatch_barrier_sync_f_invoke + 126
    	8   libdispatch.dylib                   0x0000000105f1c304 _dispatch_barrier_sync_f_slow + 529
    	9   HuginTests                          0x0000000113bccbc9 -[RCNConfigDBManager loadMainWithBundleIdentifier:fromSource:] + 249
    	10  HuginTests                          0x0000000113bc9506 -[RCNConfigContent loadConfigFromMainTable] + 70
    	11  HuginTests                          0x0000000113bc7a78 -[RCNConfigContent initWithDBManager:] + 380
    	12  HuginTests                          0x0000000113bc459e -[FIRRemoteConfig initWithDBManager:] + 156
    	13  HuginTests                          0x0000000113bc44da __31+[FIRRemoteConfig remoteConfig]_block_invoke + 86
    	14  libdispatch.dylib                   0x0000000105f33792 _dispatch_client_callout + 8
    	15  libdispatch.dylib                   0x0000000105f1ea8d dispatch_once_f + 69
    	16  HuginTests                          0x0000000113bc4481 +[FIRRemoteConfig remoteConfig] + 45
    	17  HuginTests                          0x0000000113bc3ce3 +[FIRRemoteConfig(FIRApp) receivedReadyToConfigureNotification:] + 194
    	18  CoreFoundation                      0x0000000104944c2c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    	19  CoreFoundation                      0x0000000104944b29 _CFXRegistrationPost + 425
    	20  CoreFoundation                      0x0000000104944892 ___CFXNotificationPost_block_invoke + 50
    	21  CoreFoundation                      0x0000000104908102 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1826
    	22  CoreFoundation                      0x0000000104907261 _CFXNotificationPost + 673
    	23  Foundation                          0x0000000103da7ca4 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
    	24  HuginTests                          0x0000000113b71b6e +[FIRApp sendNotificationsToSDKs:] + 381
    	25  HuginTests                          0x0000000113b70b47 +[FIRApp configureDefaultAppWithOptions:sendingNotifications:] + 302
    	26  HuginTests                          0x0000000113b707f7 +[FIRApp configure] + 381
    	27  HuginTests                          0x0000000113a76118 _TFC10HuginTests9TestSetupcfT_S0_ + 136
    	28  HuginTests                          0x0000000113a76201 _TToFC10HuginTests9TestSetupcfT_S0_ + 17
    	29  XCTest                              0x000000010433383b _XCTestMain + 508
    	30  xctest                              0x0000000103d18679 xctest + 9849
    	31  libdyld.dylib                       0x0000000105f7c65d start + 1
    	32  ???                                 0x0000000000000005 0x0 + 5
    )
    libc++abi.dylib: terminating with uncaught exception of type NSException
```

How can I fix this?

**Comment by paulb777** [comment]
@ugenlik Thanks for the report. Here are a few preliminary thoughts:

- Firebase uses dummy sample apps for its own unit testing, so the scenario you describe is not part of our current test matrix. For example, see https://github.com/firebase/firebase-ios-sdk/tree/master/Example/Core

- The crash looks related to a missing bundle - see this Stack Overflow question for context - https://stackoverflow.com/questions/38476885/set-host-application-to-none-for-test-target-then-unable-to-find-bundle-resou

- The crash looks specific to RemoteConfig. Does removing RemoteConfig from your Podfile fix the problem? Or lead to a different crash?

- Are you able to share a project that demonstrates the issue?

**Comment by ugenlik** [comment]
@paulb777 

As you can see in my code I tried to help Firebase by setting it's options to get the Test Bundle and  `GoogleService-Info.plist` to test Bundle. It didnt help.

Removing RemoteConfig from pod file solves the problem however Remote config is the only reason why I am trying to set this test case because I wanna be able to reach some local values in unit tests after Firebase configure

I'd like to be able to this basically 
```
enum ValueKey: String {
    case apiBaseUrl
    case loginUrl
}

let defaultValues = [ValueKey.apiBaseUrl.rawValue: "whatever" as NSObject]
RemoteConfig.remoteConfig().setDefaults(defaultValues)
```
then get it 

```
let optionalUrlString: String? = RemoteConfig.remoteConfig( [ValueKey.apiBaseUrl.rawValue].stringValue
```


Here is attached a sample project throws the same error when test case runs

[NoHost.zip](https://github.com/firebase/firebase-ios-sdk/files/1162801/NoHost.zip)

**Comment by chliangGoogle** [comment]
Thank you for using Firebase. We've repro your case and here's a solution:

Remote Config won't work if main bundle is missing. We will push a new release to display better error message instead of crashing.

A temporary solution is to mock the method  [NSBundle mainBundle].bundleIdentifier with a arbitrary string.

Let us know if you still run into issues.

**Comment by paulb777** [comment]
Fixed in Firebase 4.1.0, FirebaseRemoteConfig 2.0.2
